examples/ES/learner.py

```diff
@@ -0,0 +1,214 @@
+        all_noise_indices = []
+        all_training_rewards = []
+        all_training_lengths = []
+        all_eval_rewards = []
+        all_eval_lengths = []
+
+        for result in results:
+            all_eval_rewards.extend(result['eval_rewards'])
+            all_eval_lengths.extend(result['eval_lengths'])
+
+            all_noise_indices.extend(result['noise_indices'])
+            all_training_rewards.extend(result['noisy_rewards'])
+            all_training_lengths.extend(result['noisy_lengths'])
+
+        assert len(all_eval_rewards) == len(all_eval_lengths)
+        assert (len(all_noise_indices) == len(all_training_rewards) ==
+                len(all_training_lengths))
+
+        self.sample_total_episodes += num_episodes
+        self.sample_total_steps += num_timesteps
+
+        eval_rewards = np.array(all_eval_rewards)
+        eval_lengths = np.array(all_eval_lengths)
+        noise_indices = np.array(all_noise_indices)
+        noisy_rewards = np.array(all_training_rewards)
+        noisy_lengths = np.array(all_training_lengths)
+
+        # normalize rewards to (-0.5, 0.5)
+        proc_noisy_rewards = utils.compute_centered_ranks(noisy_rewards)
+        noises = [
+            self.noise.get(index, self.agent.weights_total_size)
+            for index in noise_indices
+        ]
+
+        # Update the parameters of the model.
+        self.agent.learn(proc_noisy_rewards, noises)
+        self.latest_flat_weights = self.agent.get_flat_weights()
+
+        # Update obs filter
+        self._update_filter()
+
+        # Store the evaluate rewards
+        if len(all_eval_rewards) > 0:
+            self.eval_rewards_stat.add(np.mean(eval_rewards))
+            self.eval_lengths_stat.add(np.mean(eval_lengths))
+
+        metrics = {
+            "episodes_this_iter": noisy_lengths.size,
+            "sample_total_episodes": self.sample_total_episodes,
+            'sample_total_steps': self.sample_total_steps,
+            "evaluate_rewards_mean": self.eval_rewards_stat.mean,
+            "evaluate_steps_mean": self.eval_lengths_stat.mean,
+            "timesteps_this_iter": noisy_lengths.sum(),
+        }
+
+        self.log_metrics(metrics)
+        return metrics
+
+    def _update_filter(self):
+        # Send get_filter signal to all actors
+        for q in self.actors_signal_input_queues:
+            q.put({'signal': 'get_filter'})
+
+        filters = []
+        # Collect filters from  all actors and update global filter
+        for q in self.actors_output_queues:
+            actor_filter = q.get()
+            self.obs_filter.apply_changes(actor_filter)
+
+        # Send set_filter signal to all actors
+        self.latest_obs_filter = self.obs_filter.as_serializable()
+        for q in self.actors_signal_input_queues:
+            q.put({'signal': 'set_filter'})
+
+    def log_metrics(self, metrics):
+        logger.info(metrics)
+        self.csv_logger.log_dict(metrics)
+        for k, v in metrics.items():
```

Comment:
Before adding a new data point, pls check if it is None.

Reply:
done